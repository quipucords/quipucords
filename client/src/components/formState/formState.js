import React from 'react';
import PropTypes from 'prop-types';
import _cloneDeep from 'lodash/cloneDeep';
import _isEqual from 'lodash/isEqual';
import helpers from '../../common/helpers';

const initialState = {
  isUpdating: false,
  isSubmitting: false,
  isValid: null,
  isValidating: false,
  submitCount: 0
};

class FormState extends React.Component {
  static checkedSetValues({ setValuesAssumeBoolIsChecked, setValues }) {
    const checked = {};

    if (setValuesAssumeBoolIsChecked) {
      Object.keys(setValues).forEach(key => {
        if (typeof setValues[key] === 'boolean') {
          checked[key] = setValues[key];
        }
      });
    }

    return checked;
  }

  static isPromise(obj) {
    return Object.prototype.toString.call(obj) === '[object Promise]';
  }

  constructor(props) {
    super(props);

    this.touched = {};
    this.checked = FormState.checkedSetValues(props);

    this.refUpdating = false;
    this.refValues =
      props.resetUsingSetValues === true || props.setValuesOnUpdate === true ? _cloneDeep(props.setValues) : null;

    this.errors = {};
    this.values = _cloneDeep(props.setValues);

    this.state = {
      ...initialState
    };
  }

  componentDidMount() {
    const { validateOnMount } = this.props;

    if (validateOnMount === true) {
      this.validateOnMount();
    }
  }

  componentDidUpdate() {
    const { refValues } = this;
    const { setValuesOnUpdate, setValues } = this.props;

    if (setValuesOnUpdate === true && !_isEqual(refValues, setValues)) {
      this.updateComponentValues();
    }
  }

  onEventCustom = custom => {
    const eventArray = (Array.isArray(custom) && custom) || (custom && [custom]);

    if (!eventArray.length) {
      return;
    }

    eventArray
      .filter(event => 'name' in event && ('value' in event || 'checked' in event))
      .forEach(event => this.onEvent({ target: { ...event }, persist: helpers.noop, type: 'custom' }));
  };

  onEvent = event => {
    const { touched, values } = this;
    const { id, name, value, checked } = event.options ? { ...event } : event.target;

    event.persist();

    const targetName = name || id || 'generated form state target, add name or id attr to field';

    this.touched = { ...touched, [targetName]: true };
    this.values = { ...values, [targetName]: value };

    if (checked !== undefined) {
      this.checked = { ...this.checked, [targetName]: checked };
    }

    this.setState(
      {
        isUpdating: true,
        isValidating: true
      },
      () =>
        this.validate(event).then(updatedErrors => {
          const setUpdateErrors = { ...((updatedErrors && updatedErrors[0]) || updatedErrors || {}) };
          const checkIsValid = !Object.keys(setUpdateErrors).length;

          this.errors = setUpdateErrors;

          this.setState({
            isUpdating: false,
            isValid: checkIsValid,
            isValidating: false
          });
        })
    );
  };

  onReset = event => {
    const { refValues, values } = this;
    const { setValuesAssumeBoolIsChecked, onReset, resetUsingSetValues } = this.props;

    event.persist();

    const isResetWithSetValues = refValues && resetUsingSetValues === true;
    const updatedValues = (isResetWithSetValues && _cloneDeep(refValues)) || {};
    const updatedChecked =
      (isResetWithSetValues && FormState.checkedSetValues(setValuesAssumeBoolIsChecked, updatedValues)) || {};

    this.values = updatedValues;
    this.checked = updatedChecked;
    this.errors = {};
    this.touched = {};

    this.setState({
      ...initialState
    });

    if (isResetWithSetValues) {
      onReset({ event, ..._cloneDeep({ values: updatedValues, prevValues: values }) });
    } else {
      // Resetting the values, potentially, will throw the controlled vs uncontrolled messaging.
      onReset({ event, values: {}, ..._cloneDeep({ prevValues: values }) });
    }
  };

  onSubmit = event => {
    const { submitCount } = this.state;

    event.persist();
    event.preventDefault();

    this.setState(
      {
        submitCount: submitCount + 1,
        isSubmitting: true,
        isUpdating: true,
        isValidating: true
      },
      () =>
        this.validate(event).then(updatedErrors => {
          const setUpdateErrors = { ...((updatedErrors && updatedErrors[0]) || updatedErrors || {}) };
          const checkIsValid = !Object.keys(setUpdateErrors).length;

          this.errors = setUpdateErrors;
          this.touched = {};

          this.setState(
            {
              isValid: checkIsValid,
              isValidating: false
            },
            () =>
              checkIsValid &&
              this.submit(event).then(() => {
                this.setState({
                  isSubmitting: false,
                  isUpdating: false
                });
              })
          );
        })
    );
  };

  submit(event = { type: 'submit' }) {
    const { checked, errors, values, touched } = this;
    const { onSubmit } = this.props;

    const checkPromise = onSubmit({
      event,
      ..._cloneDeep({ ...this.state, checked, errors, values, touched })
    });

    if (FormState.isPromise(checkPromise)) {
      return checkPromise;
    }

    return {
      then: callback => callback()
    };
  }

  validate(event = { type: 'validate' }) {
    const { checked, errors, values, touched } = this;
    const { validate } = this.props;

    const checkPromise = validate({
      event,
      ..._cloneDeep({ ...this.state, checked, errors, values, touched })
    });

    if (FormState.isPromise(checkPromise)) {
      return checkPromise;
    }

    return {
      then: callback => callback(checkPromise)
    };
  }

  validateOnMount(event = { type: 'mount' }) {
    this.validateOnUpdate(event);
  }

  validateOnUpdate(event = { type: 'update' }) {
    this.setState(
      {
        isUpdating: true,
        isValidating: true
      },
      () =>
        this.validate(event).then(updatedErrors => {
          const setUpdateErrors = { ...((updatedErrors && updatedErrors[0]) || updatedErrors || {}) };
          const checkIsValid = !Object.keys(setUpdateErrors).length;

          this.errors = setUpdateErrors;

          this.setState({
            isUpdating: false,
            isValidating: false,
            isValid: checkIsValid
          });
        })
    );
  }

  updateComponentValues() {
    const { setValues, setValuesAssumeBoolIsChecked, validateOnUpdate } = this.props;

    this.setState(
      {
        isUpdating: true
      },
      () => {
        this.checked = FormState.checkedSetValues(setValuesAssumeBoolIsChecked, setValues);
        this.refValues = _cloneDeep(setValues);
        this.values = _cloneDeep(setValues);

        if (validateOnUpdate) {
          this.validateOnUpdate();
        } else {
          this.setState({
            isUpdating: false
          });
        }
      }
    );
  }

  render() {
    const { checked, errors, values, touched } = this;
    const { children } = this.props;

    return (
      <React.Fragment>
        {children({
          handleOnEventCustom: this.onEventCustom,
          handleOnEvent: this.onEvent,
          handleOnReset: this.onReset,
          handleOnSubmit: this.onSubmit,
          ..._cloneDeep({ ...this.state, checked, errors, values, touched })
        })}
      </React.Fragment>
    );
  }
}

FormState.propTypes = {
  children: PropTypes.func.isRequired,
  onReset: PropTypes.func,
  onSubmit: PropTypes.func,
  resetUsingSetValues: PropTypes.bool,
  setValues: PropTypes.object,
  setValuesOnUpdate: PropTypes.bool,
  setValuesAssumeBoolIsChecked: PropTypes.bool,
  validate: PropTypes.func,
  validateOnMount: PropTypes.bool,
  validateOnUpdate: PropTypes.bool
};

FormState.defaultProps = {
  onReset: helpers.noop,
  onSubmit: helpers.noop,
  resetUsingSetValues: true,
  setValues: {},
  setValuesOnUpdate: false,
  setValuesAssumeBoolIsChecked: true,
  validate: helpers.noop,
  validateOnMount: false,
  validateOnUpdate: false
};

export { FormState as default, FormState };
