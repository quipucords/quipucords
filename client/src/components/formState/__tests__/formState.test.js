import React from 'react';
import { mount } from 'enzyme';
import { FormState } from '../formState';

describe('FormState Component', () => {
  it('should render a basic component', () => {
    const props = { setValues: { lorem: 'ipsum' } };

    const component = mount(
      <FormState {...props}>
        {({ values }) => (
          <form>
            <label>
              Lorem
              <input name="lorem" value={values.lorem} readOnly type="text" />
            </label>
          </form>
        )}
      </FormState>
    );

    const componentInstance = component.instance();
    expect(componentInstance.props).toMatchSnapshot('initial props');
    expect(componentInstance.state).toMatchSnapshot('initial state');

    expect(component.render()).toMatchSnapshot('basic render');
  });

  it('should update handle reset, changes, and submit events while updating state', () => {
    const props = { setValues: { lorem: 'ipsum' }, resetUsingSetValues: true, validate: () => ({}) };

    const component = mount(
      <FormState {...props}>
        {({ values, handleOnEvent, handleOnReset, handleOnSubmit }) => (
          <form onSubmit={handleOnSubmit} onReset={handleOnReset}>
            <label>
              Lorem
              <input name="lorem" value={values.lorem} type="text" onChange={handleOnEvent} />
            </label>
            <button type="submit">Submit</button>
          </form>
        )}
      </FormState>
    );

    const componentInstance = component.instance();

    componentInstance.onEvent({
      target: { value: 'dolor', name: 'lorem' },
      persist: () => {},
      preventDefault: () => {}
    });
    expect(component.state()).toMatchSnapshot('onevent');
    expect(componentInstance.values).toMatchSnapshot('onevent values updated');

    componentInstance.onReset({ persist: () => {} });
    expect(component.state()).toMatchSnapshot('onreset');
    expect(componentInstance.values).toMatchSnapshot('reset values updated');

    componentInstance.onSubmit({ persist: () => {}, preventDefault: () => {} });
    expect(component.state()).toMatchSnapshot('onsubmit');
  });

  it('should do a basic validation check', () => {
    const props = {
      setValues: {
        lorem: 'ipsum'
      },
      validate: ({ values }) => {
        const updatedErrors = {};

        if (!values.lorem) {
          updatedErrors.lorem = 'required';
        }

        return updatedErrors;
      }
    };

    const component = mount(
      <FormState {...props}>
        {({ errors, values, handleOnEvent }) => (
          <form>
            <label>
              Lorem
              <input id="lorem" value={values.lorem} type="text" onChange={handleOnEvent} />
              <span className="error">{errors.lorem}</span>
            </label>
          </form>
        )}
      </FormState>
    );

    const componentInstance = component.instance();
    expect(componentInstance.errors).toMatchSnapshot('initial errors');

    const mockEvent = { target: { value: '', id: 'lorem' }, persist: () => {}, preventDefault: () => {} };
    componentInstance.onEvent(mockEvent);
    expect(componentInstance.errors).toMatchSnapshot('basic validation');
  });

  it('should handle custom events', () => {
    const props = {
      setValues: {
        lorem: 'ipsum',
        dolor: ''
      },
      validate: ({ values }) => {
        const updatedErrors = {};

        if (!values.lorem) {
          updatedErrors.lorem = 'required';
        }

        return updatedErrors;
      }
    };

    const component = mount(
      <FormState {...props}>
        {({ errors, values, handleOnEventCustom }) => (
          <form>
            <label>
              Lorem
              <input name="lorem" value={values.lorem} type="text" onChange={handleOnEventCustom} />
              <span className="error">{errors.lorem}</span>
              <input name="dolor" value={values.dolor} type="hidden" />
            </label>
          </form>
        )}
      </FormState>
    );

    const componentInstance = component.instance();
    componentInstance.onEventCustom({ name: 'lorem', value: 'woot' });
    expect(componentInstance.values).toMatchSnapshot('single custom event');

    componentInstance.onEventCustom([{ name: 'lorem', value: 'woot again' }, { name: 'dolor', value: 'sit' }]);
    expect(componentInstance.values).toMatchSnapshot('multiple custom events');
  });

  it('should clone returned values to avoid mutation by consumer', () => {
    const props = {
      setValues: {
        lorem: 'ipsum'
      },
      validate: ({ values }) => {
        const updatedErrors = {};

        // eslint-disable-next-line
        values.lorem = 'mutated';

        return updatedErrors;
      }
    };

    const component = mount(
      <FormState validateOnmount {...props}>
        {({ values }) => <div>Lorem = {values.lorem}</div>}
      </FormState>
    );

    expect(component.instance().values).toMatchSnapshot('not mutated');
  });
});
