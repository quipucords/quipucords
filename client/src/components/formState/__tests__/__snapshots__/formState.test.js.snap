// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic render 1`] = `
<form>
  <label>
    Lorem
    <input
      name="lorem"
      readonly=""
      type="text"
      value="ipsum"
    />
  </label>
</form>
`;

exports[`basic validation 1`] = `
Object {
  "lorem": "required",
}
`;

exports[`initial errors 1`] = `Object {}`;

exports[`initial props 1`] = `
Object {
  "children": [Function],
  "onReset": [Function],
  "onSubmit": [Function],
  "resetUsingSetValues": true,
  "setValues": Object {
    "lorem": "ipsum",
  },
  "setValuesAssumeBoolIsChecked": true,
  "setValuesOnUpdate": false,
  "validate": [Function],
  "validateOnMount": false,
  "validateOnUpdate": false,
}
`;

exports[`initial state 1`] = `
Object {
  "isSubmitting": false,
  "isUpdating": false,
  "isValid": null,
  "isValidating": false,
  "submitCount": 0,
}
`;

exports[`multiple custom events 1`] = `
Object {
  "dolor": "sit",
  "lorem": "woot again",
}
`;

exports[`not mutated 1`] = `
Object {
  "lorem": "ipsum",
}
`;

exports[`onevent 1`] = `
Object {
  "isSubmitting": false,
  "isUpdating": false,
  "isValid": true,
  "isValidating": false,
  "submitCount": 0,
}
`;

exports[`onevent values updated 1`] = `
Object {
  "lorem": "dolor",
}
`;

exports[`onreset 1`] = `
Object {
  "isSubmitting": false,
  "isUpdating": false,
  "isValid": null,
  "isValidating": false,
  "submitCount": 0,
}
`;

exports[`onsubmit 1`] = `
Object {
  "isSubmitting": false,
  "isUpdating": false,
  "isValid": true,
  "isValidating": false,
  "submitCount": 1,
}
`;

exports[`reset values updated 1`] = `
Object {
  "lorem": "ipsum",
}
`;

exports[`single custom event 1`] = `
Object {
  "dolor": "",
  "lorem": "woot",
}
`;
